### PR TITLE
feat: fees v3

### PR DESCRIPTION
**Pull request by tynes**
**Description**
Updates the fee calculation to match https://github.com/ethereum-optimism/optimism/issues/823#issuecomment-851760512
This scales down the L2 Gas Limit such that it has the granularity of 10,000. The numbers should now show correctly in metamask

### CODE REVIEW

File: l2geth/rollup/fees/rollup_fee.go [hunk 1]
```diff
@@ -52,20 +54,31 @@ var bigHundredMillion = new(big.Int).SetUint64(hundredMillion)
...
 	l1Fee := new(big.Int).Mul(l1GasPrice, l1GasLimit)
-	l2Fee := new(big.Int).Mul(l2GasPrice, l2GasLimit)
+	l2Fee := new(big.Int).Mul(l2GasPrice, roundedL2GasLimit)
 	sum := new(big.Int).Add(l1Fee, l2Fee)
 	scaled := new(big.Int).Div(sum, bigFeeScalar)
-	remainder := new(big.Int).Mod(scaled, bigHundredMillion)
-	scaledSum := new(big.Int).Add(scaled, bigHundredMillion)
-	rounded := new(big.Int).Sub(scaledSum, remainder)
-	result := new(big.Int).Add(rounded, l2GasLimit)
+	rounded := Ceilmod(scaled, bigHundredMillion)
```
Comment (K-Ho) [inline l2geth/rollup/fees/rollup_fee.go:62]:
```suggestion
	rounded := Ceilmod(scaled, BigTenThousand)
```

File: l2geth/rollup/fees/rollup_fee.go [hunk 2]
```diff
@@ -27,6 +28,7 @@ const TxGasPrice uint64 = feeScalar + (feeScalar / 2)
 var BigTxGasPrice = new(big.Int).SetUint64(TxGasPrice)
 var bigFeeScalar = new(big.Int).SetUint64(feeScalar)
```
Comment (K-Ho) [inline l2geth/rollup/fees/rollup_fee.go:29]: `const feeScalar uint64 = 10_000_000`

File: packages/core-utils/src/fees.ts [hunk 3]
```diff
@@ -32,28 +33,45 @@ function encode(input: EncodableL2GasLimit): BigNumber {
     l2GasPrice = BigNumber.from(l2GasPrice)
   }
   const l1GasLimit = calculateL1GasLimit(data)
+  const roundedL2GasLimit = ceilmod(l2GasLimit, tenThousand)
   const l1Fee = l1GasLimit.mul(l1GasPrice)
-  const l2Fee = l2GasLimit.mul(l2GasPrice)
+  const l2Fee = roundedL2GasLimit.mul(l2GasPrice)
   const sum = l1Fee.add(l2Fee)
   const scaled = sum.div(feeScalar)
```
Comment (K-Ho) [inline packages/core-utils/src/fees.ts:40]: feeScalar must be `10_000_000`

File: packages/core-utils/src/fees.ts [hunk 4]
```diff
@@ -32,28 +33,45 @@ function encode(input: EncodableL2GasLimit): BigNumber {
...
-  const remainder = scaled.mod(hundredMillion)
-  const scaledSum = scaled.add(hundredMillion)
-  const rounded = scaledSum.sub(remainder)
-  return rounded.add(l2GasLimit)
+  const rounded = ceilmod(scaled, hundredMillion)
```
Comment (K-Ho) [inline packages/core-utils/src/fees.ts:41]:
```suggestion
  const rounded = ceilmod(scaled, tenThousand)
```

**Review by K-Ho — CHANGES_REQUESTED** [review]
couple of changes - need to scale the `tx.gasPrice` up `10,000x` and then scale the `scaled` (fee) down `10,000x`

File: l2geth/core/tx_pool.go [hunk 5]
```diff
@@ -540,10 +539,14 @@ func (pool *TxPool) validateTx(tx *types.Transaction, local bool) error {
 	}
 
 	// Ensure the transaction doesn't exceed the current block limit gas.
```
Comment (tynes) [inline l2geth/core/tx_pool.go:541]: Ideally we no longer use this codepath at all so we can reduce the diff here

**Review by K-Ho — APPROVED** [review]
LGTM

**Pull request merged**